### DISCUSSION TITLE
Node version for Frontend; How to run the backend?

### DISCUSSION THREAD

**Discussion by vikrantuk** [nammayatri/nammayatri#264] [Q&A] [answered]
Hi there!

I just wanted to take a moment to express my appreciation for your awesome repository. Your code is incredibly well-organized, easy to understand, and has been a great resource for me as a developer. Thank you for all the hard work you've put into it!

I do have a quick question for you. I was wondering if you could let me know which version(s) of Node.js your repository supports? I want to make sure that I have the correct version installed before I start working with your code.

Thanks again for all your hard work, and I look forward to hearing back from you soon!

**Comment by witcher-shailesh** [comment, marked as answer]
Hi @vikrantuk, 

Thanks a lot for your contribution and kind words. Will be looking forward for more contribution from your end. For your query, we are supporting node.js version V_14.17.6

Happy Coding!!

**Reply by vikrantuk** [reply]
Hi @witcher-shailesh, thank you for your response.

I tried setting up backend repo too and while running '**stack build --fast**' I'm running into **ghc panic bad argument error**. Pasting the full log below:

`vik@vik-VirtualBox:~/nammayatri-15.6.0/Backend$ stack build --fast
DEPRECATED: The package at Repo from https://github.com/nammayatri/beckn-gateway.git, commit 7094d2af08055f37ca01f2891d19f3153e619be1 in subdir app/gateway does not include a cabal file.
Instead, it includes an hpack package.yaml file for generating a cabal file.
This usage is deprecated; please see https://github.com/commercialhaskell/stack/issues/5210.
Support for this workflow will be removed in the future.

DEPRECATED: The package at Repo from https://github.com/nammayatri/beckn-gateway.git, commit 7094d2af08055f37ca01f2891d19f3153e619be1 in subdir app/mock-registry does not include a cabal file.
Instead, it includes an hpack package.yaml file for generating a cabal file.
This usage is deprecated; please see https://github.com/commercialhaskell/stack/issues/5210.
Support for this workflow will be removed in the future.

ghc: panic! (the 'impossible' happened)
  (GHC version 8.8.4 for x86_64-unknown-linux):
	Prelude.chr: **bad argument: 2634022915**

Please report this as a GHC bug:  https://www.haskell.org/ghc/reportabug


Error: [S-6374]
       While building simple Setup.hs (scroll up to its section to see the
       error) using:
       /home/vik/.stack/programs/x86_64-linux/ghc-tinfo6-8.8.4/bin/ghc-8.8.4 -rtsopts -threaded -clear-package-db -global-package-db -hide-all-packages -package base -main-is StackSetupShim.mainOverride -package Cabal-3.0.1.0 /home/vik/.stack/setup-exe-src/setup-SvXsv1f_.hs /home/vik/.stack/setup-exe-src/setup-shim-SvXsv1f_.hs -o /home/vik/.stack/setup-exe-cache/x86_64-linux-tinfo6/tmp-Cabal-simple_SvXsv1f__3.0.1.0_ghc-8.8.4
       Process exited with code: ExitFailure 1 `

Can you please provide some insight over this too?
Installed Ghc version is 9.4.4 and stack version is 2.9.3

Thank you

**Reply by Magizhan** [reply]
@vikrantuk  We are making the setup simple with nix and will update the steps soon. 
Haskell version 8.8.4 is needed for now. 

@piyushKumar-1  Can you please help @vikrantuk to set up Backend / update the setup steps ? 
@hkmangla When will we be able to update nix set up steps ?

**Reply by srid** [reply]
@vikrantuk I'll wait for somebody else to respond to this, but if you have [Nix installed](https://haskell.flake.page/nix), you can also run the following command to get the backend built:

```sh
nix run nixpkgs#cachix use nammayatri
nix build .#nammayatri
ls result/
```

We'll soon be switching over from Stack to Nix which provides various advantages like dealing with "works on my machine, but not their's" issues like the above.

**Reply by hkmangla** [reply]
@srid As we have merged the nix to main branch. We can remove stack set up from Github README and move to full time nix for new contributors. We can keep stack for 2 more weeks for old contributors to who are still taking up nix.

@vikrantuk Please set up using nix, we will update the README too.

**Reply by vikrantuk** [reply]
> @vikrantuk I'll wait for somebody else to respond to this, but if you have [Nix installed](https://haskell.flake.page/nix), you can also run the following command to get the backend built:
> 
> ```shell
> nix run nixpkgs#cachix use nammayatri
> nix build .#nammayatri
> ls result/
> ```
> 
> We'll soon be switching over from Stack to Nix which provides various advantages like dealing with "works on my machine, but not their's" issues like the above.

Hi @srid, I was able to build repo with nix. But I'm a complete newbie and have no idea on how to run the built code using nix. Can you please provide further guidance on this please?

**Reply by srid** [reply]
@vikrantuk We are going to use Nix throughout the dev workflow; see https://github.com/nammayatri/nammayatri/issues/356

Meanwhile, things have to be done manually. After building with Nix, do you see the binaries under the `result` symlink? On my machine, I see:

```
❯ ls result/bin/
allocation-service-exe           mock-google-exe                              rider-dashboard-exe
beckn-cli-exe                    mock-idfy-exe                                route-extractor-exe
driver-offer-allocator-exe       mock-public-transport-provider-platform-exe  scheduler-example-app-exe
driver-tracking-healthcheck-exe  mock-rider-platform-exe                      scheduler-example-scheduler-exe
dynamic-offer-driver-app-exe     mock-sms-exe                                 search-result-aggregator-exe
example-service-exe              provider-dashboard-exe                       static-offer-driver-app-exe
image-api-helper-exe             public-transport-rider-platform-exe          transporter-scheduler-exe
kafka-consumers-exe              public-transport-search-consumer-exe
mock-fcm-exe                     rider-app-exe
```

`Backend/README.md` has some instructions on running the backend. But it doesn't seem complete (cc @hkmangla). In particular, you may also want to run `make run-mobility-stack`, but to make `run.sh` use the Nix binaries you may have to apply this patch, and tweak accordingly:

```diff
diff --git a/Backend/dev/run.sh b/Backend/dev/run.sh
index 22b969ba..403b6223 100755
--- a/Backend/dev/run.sh
+++ b/Backend/dev/run.sh
@@ -31,7 +31,7 @@ map () { while read -r line; do $1 "$line"; done; }
 
 to_log_filename () { echo "$1.log"; }
 
-to_run_cmd () { echo "stack exec $1-exe 2>&1 | tee $(to_log_filename $1)"; }
+to_run_cmd () { echo "result/bin/$1-exe 2>&1 | tee $(to_log_filename $1)"; }
 
 rm -f $(apps | map to_log_filename)
 apps | map to_run_cmd | parallel -j $(apps | wc -l)
```

I realize this is all unnecessarily complicated, which is why we are replacing Stack with Nix. All of the above will be obviated and automated as part of #356 - after which you can run the backend using a single command.

**Reply by vikrantuk** [reply]
> ```shell
> nix build .#nammayatri
> ```

@srid I tried running
`make run-mobility-stack`
but it results in multiple missing files errors:
`vik@vik-VirtualBox:~/nammayatri/Backend$ make run-mobility-stack 
# Run all binaries needed for mobility network
./dev/run.sh
/usr/bin/bash: result/bin/beckn-gateway-exe: No such file or directory
driver-tracking-healthcheck-exe: 
 ./dhall-configs/dev/dynamic-offer-driver-app.dhall
   ./dhall-configs/dev/common.dhall
     ./dhall-configs/dev/secrets/common.dhall
       ./dhall-configs/dev/secrets/top-secret.dhall

Error: Missing file ./dhall-configs/dev/secrets/top-secret.dhall

1                 ./top-secret.dhall
`

**Reply by srid** [reply]
@hkmangla Do you know what should be done about the `.dhall` configuration ?

@vikrantuk I've nixified `dev/run.sh` in a separate branch. Run `nix run github:nammayatri/nammayatri/nixdev#run-nammayatri` (instead of `dev/run.sh`). This is still WIP as part of https://github.com/nammayatri/nammayatri/issues/356

**Reply by vikrantuk** [reply]
> @hkmangla Do you know what should be done about the `.dhall` configuration ?
> 
> @vikrantuk I've nixified `dev/run.sh` in a separate branch. Run `nix run github:nammayatri/nammayatri/nixdev#run-nammayatri` (instead of `dev/run.sh`). This is still WIP as part of #356

Hi @srid thank you for responding. I made a little progress by running `nix run github:nammayatri/nammayatri/nixdev#run-nammayatri`

But most of the services are still failing for `Error: Missing file ./dhall-configs/dev/secrets/top-secret.dhall` with exit code 1.

`scheduler-example-0.1.0.0` is failing for `bash: /nix/store/bvcm9nfdfkk1xxbrgji9gr5w48ip9hxh-scheduler-example-0.1.0.0/bin/scheduler-example: No such file or directory` with exit code 127

Thank you for your patience.

**Reply by hkmangla** [reply]
@vikrantuk Just add this file in the place it has asked for. There is already a template file, copy the content from it.

**Reply by vikrantuk** [reply]
> @vikrantuk Just add this file in the place it has asked for. There is already a template file, copy the content from it.

Hi @hkmangla, I created top-secret.dhall with valid key values. However, I'm still having trouble getting the platform started. Now I'm getting `provider-dashboard-exe: Network.Socket.connect: <socket: 16>: does not exist (Connection refused)`
 and  
`bash: /nix/store/bvcm9nfdfkk1xxbrgji9gr5w48ip9hxh-scheduler-example-0.1.0.0/bin/scheduler-example: No such file or directory` with exit code 127 still exist

**Reply by srid** [reply]
> scheduler-example-0.1.0.0 is failing for `bash: /nix/store/bvcm9nfdfkk1xxbrgji9gr5w48ip9hxh-scheduler-example-0.1.0.0/bin/scheduler-example: No such file or directory with exit code 127`

I just fixed this; could you try again, @vikrantuk ? `nix run --refresh github:nammayatri/nammayatri/nixdev#run-nammayatri`

> `provider-dashboard-exe: Network.Socket.connect: <socket: 16>: does not exist (Connection refused)`

Try this in a different terminal:

```
git clone https://github.com/nammayatri/nammayatri
cd nammayatri
git checkout nixdev
cd Backend
nix develop -c , backend-run-svc  # Needs docker installed
````

And then run `nix run --refresh github:nammayatri/nammayatri/nixdev#run-nammayatri` in a separate terminal.

With this, I got all processes running, except `driver-offer-allocator` and `static-offer-driver-app-allocator`.

<img width="1131" alt="Screenshot 2023-03-29 at 10 34 06 AM" src="https://user-images.githubusercontent.com/3998/228573308-c9065584-6c93-46c5-8d9c-fe25668558fd.png">

**Reply by vikrantuk** [reply]
> Try this in a different terminal:
> 
> ```
> git clone https://github.com/nammayatri/nammayatri
> cd nammayatri
> git checkout nixdev
> cd Backend
> nix develop -c , backend-run-svc  # Needs docker installed
> ```
> 
> And then run `nix run --refresh github:nammayatri/nammayatri/nixdev#run-nammayatri` in a separate terminal.
> 
> With this, I got all processes running, except `driver-offer-allocator` and `static-offer-driver-app-allocator`.

Thanks @srid, I was finally able to start the backend huge thank you to you :)
I was able to run `driver-offer-allocator` and `static-offer-driver-app-allocator` by starting them again.

**Reply by srid** [reply]
FYI: The command now is:

```
nix run github:nammayatri/nammayatri/nixdev#run-mobility-stack
```

**Reply by vikrantuk** [reply]
> FYI: The command now is:
> 
> ```
> nix run github:nammayatri/nammayatri/nixdev#run-mobility-stack
> ```

Thanks @srid, that simplifies the backend setup process. Do you have similar simpler process for running frontend too? I went through readme of Frontend and it is bit confusing for newbies.

**Reply by srid** [reply]
> Do you have similar simpler process for running frontend too?

Not right now, but we definitely plan to nixify the frontend as a next step. After that you can get the full-stack nammaytri running with one (`nix run ..`) command!

**Reply by vikrantuk** [reply]
> > Do you have similar simpler process for running frontend too?
> 
> Not right now, but we definitely plan to nixify the frontend as a next step. After that you can get the full-stack nammaytri running with one (`nix run ..`) command!

@srid That will be wonderful. Will you be able to help in setting up frontend too?

**Reply by kishorgandham** [reply]
@srid what version of nix are you using?

**Reply by srid** [reply]
`2.13.3`

If you are installing Nix, I recommend https://github.com/DeterminateSystems/nix-installer#the-determinate-nix-installer

**Reply by srid** [reply]
> Will you be able to help in setting up frontend too?

@vikrantuk I suggest posting a new discussion topic for this. So the frontend team has an opportunity to respond.

**Reply by kishorgandham** [reply]
> `2.13.3`
> 
> If you are installing Nix, I recommend https://github.com/DeterminateSystems/nix-installer#the-determinate-nix-installer

Hi @srid thanks for this info. It was helpful. I am able to bring up the backend applications. One question on application to run on port 8019. It looks like there are several references in the code to a mock Google maps application that runs on port 8019. However, with nix, when I started the app, I did not see any app on port 8019. How do we bring up this app on port 8019?

**Reply by srid** [reply]
@kishorgandham Thanks for catching that, I just pushed https://github.com/nammayatri/nammayatri/commit/4955769bd0cebe4909d84065a1ccefcd18a8faef - try again.

**Reply by vikrantuk** [reply]
Hi @srid, I was trying to setup Backend in aws ec2 ubuntu instance and facing issues with connection to port 5434. Pasting screenshot for the reference
![image](https://user-images.githubusercontent.com/14870157/230943274-70026606-1dad-493b-9b87-2af3396793f3.png)

**Reply by srid** [reply]
You want to run the docker containers as well, via `nix run github:nammayatri/nammayatri/nixdev#arion up` (this includes postgres).

I'm in the process of documenting this all in https://github.com/nammayatri/nammayatri/tree/nixdev/Backend

**Reply by vikrantuk** [reply]
> You want to run the docker containers as well, via `nix run github:nammayatri/nammayatri/nixdev#arion up` (this includes postgres).
> 
> I'm in the process of documenting this all in https://github.com/nammayatri/nammayatri/tree/nixdev/Backend

Oh Ok. I thought running `, backend-run-svc` takes care of starting all the docker containers

**Reply by srid** [reply]
It should do the same, if you are running it in `nix develop` shell, and assuming those containers fully started and didn't error out.

**Reply by vikrantuk** [reply]
@srid Yes I did run in nix shell and didn't see any error.
![image](https://user-images.githubusercontent.com/14870157/230946007-e1f89772-7d9e-454b-8433-b7b12016e098.png)

**Reply by srid** [reply]
It is `, run-svc` now, btw. And it starts containers in foreground mode, streaming the logs along the way.

`, run-svc -d` will run them in daemon mode.

You can also run `arion ps` or `arion logs` to inspect these containers.

**Comment by maajidz** [comment]
Can anyone kindly update the docs for both front end and backend? They seem to be outdated

**Reply by srid** [reply]
The [backend docs](https://github.com/nammayatri/nammayatri/tree/main/Backend#getting-started) should definitely be up to date. Are you having any specific issues?

**Reply by maajidz** [reply]
Yes it did work with the release verion 18.1.0, but the front end docs are outdated.  It seemd like 'CreateJsafiles' step should not be there according to the commit history.

**Reply by maajidz** [reply]
Also, we were trying to run the backend prod server, but apparently only the mock services are being, used, Is there any flag that changes the server behaviour to not use mock services, and rather use the prod version?
